gowin/pack.cc

```diff
@@ -891,13 +895,142 @@ static void pack_iologic(Context *ctx)
                     ci->addInput(id_XXX_VCC);
                     ci->connectPort(id_XXX_VCC, ctx->nets[ctx->id("$PACKER_VCC_NET")].get());
                 }
-                if (ctx->gw1n9_quirk && iob_bel != q0_dst->attrs.end()) {
-                    bool have_XXX_VSS0 =
-                            ctx->bels[ctx->getBelByNameStr(iob_bel->second.as_string())].pins.count(id_XXX_VSS0);
-                    if (have_XXX_VSS0) {
+                if (iob_bel != q0_dst->attrs.end()) {
+                    IdString io_bel_name = ctx->getBelByNameStr(iob_bel->second.as_string());
+                    if (ctx->gw1n9_quirk && ctx->bels[io_bel_name].pins.count(id_XXX_VSS0)) {
                         q0_dst->disconnectPort(id_XXX_VSS0);
                         q0_dst->connectPort(id_XXX_VSS0, ctx->nets[ctx->id("$PACKER_VCC_NET")].get());
                     }
+                    if (ctx->bels[io_bel_name].pins.count(id_XXX_1)) {
+                        q0_dst->disconnectPort(id_XXX_1);
+                        q0_dst->connectPort(id_XXX_1, ctx->nets[ctx->id("$PACKER_VCC_NET")].get());
+                    }
+                }
+            } break;
+            case ID_OSER4:  /* fall-through */
+            case ID_OSER8:  /* fall-through */
+            case ID_OSER10: /* fall-through */
+            case ID_OVIDEO: {
+                IdString output = id_Q;
+                IdString output_1 = IdString();
+                if (ci->type == id_OSER4 || ci->type == id_OSER8) {
+                    output = id_Q0;
+                    output_1 = id_Q1;
+                }
+                q0_dst = net_only_drives(ctx, ci->ports.at(output).net, is_iob, id_I);
+                NPNR_ASSERT(q0_dst != nullptr);
+
+                auto iob_bel = q0_dst->attrs.find(id_BEL);
+                if (iob_bel == q0_dst->attrs.end()) {
+                    log_info("No constraints for %s\n", ctx->nameOf(q0_dst));
+                    NPNR_ASSERT_FALSE("The pins for IDES/OSER must be specified explicitly.");
+                }
+
+                Loc loc = ctx->getBelLocation(ctx->getBelByNameStr(iob_bel->second.as_string()));
+                loc.z += BelZ::iologic_z;
+                ci->setAttr(id_BEL, ctx->getBelName(ctx->getBelByLocation(loc)).str(ctx));
+                BelId bel = ctx->getBelByLocation(loc);
+                if (bel == BelId()) {
+                    log_info("No bel for %s at %s\n", ctx->nameOf(ci), iob_bel->second.as_string().c_str());
+                    NPNR_ASSERT_FALSE("Can't place IDES/OSER here");
```

Comment:
this should probably be a nicer combined `log_error` error message rather than a `log_info` followed by assertion if it's something a user might trigger accidentally (like trying to use OSER at an IO not capable of it).